File: src/lib/timeFormatter.ts
/**
 * Utility functions for relative and exact date/time formatting.
 * Preserves backend UTC timestamp as source of truth and converts to local browser timezone.
 */

export function formatRelativeTime(dateInput?: string | Date | null): string {
  if (!dateInput) return "Recently";

  const date = typeof dateInput === "string" ? new Date(dateInput) : dateInput;
  if (isNaN(date.getTime())) return "Recently";

  const now = new Date();
  const diffMs = now.getTime() - date.getTime();
  
  // Future dates safeguard
  if (diffMs < 0) return "Just now";

  const diffSecs = Math.floor(diffMs / 1000);
  const diffMins = Math.floor(diffSecs / 60);
  const diffHours = Math.floor(diffMins / 60);
  const diffDays = Math.floor(diffHours / 24);
  const diffWeeks = Math.floor(diffDays / 7);
  const diffMonths = Math.floor(diffDays / 30);

  if (diffSecs < 60) {
    return "Just now";
  }
  if (diffMins < 60) {
    return `${diffMins}m ago`;
  }
  if (diffHours < 24) {
    return `${diffHours}h ago`;
  }
  if (diffDays === 1) {
    return "Yesterday";
  }
  if (diffDays < 14) {
    return `${diffDays}d ago`;
  }
  if (diffWeeks < 8) {
    return `${diffWeeks}w ago`;
  }
  if (diffMonths < 12) {
    return `${diffMonths}mo ago`;
  }

  // Older than 1 year -> Show actual date
  return `Posted ${date.toLocaleDateString("en-US", {
    month: "short",
    day: "numeric",
    year: "numeric",
  })}`;
}

export function formatExactDateTime(dateInput?: string | Date | null): string {
  if (!dateInput) return "Posted recently";

  const date = typeof dateInput === "string" ? new Date(dateInput) : dateInput;
  if (isNaN(date.getTime())) return "Posted recently";

  const dateStr = date.toLocaleDateString("en-US", {
    month: "long",
    day: "numeric",
    year: "numeric",
  });

  const timeStr = date.toLocaleTimeString("en-US", {
    hour: "numeric",
    minute: "2-digit",
    hour12: true,
  });

  return `Posted ${dateStr} at ${timeStr}`;
}

File: mobile/lib/core/utils/time_formatter.dart
import 'package:intl/intl.dart';

class TimeFormatter {
  /// Formats UTC ISO string or DateTime to modern relative time (e.g. "Just now", "5m ago", "2h ago", "Yesterday", "3d ago", "Posted Aug 24, 2026").
  static String formatRelativeTime(dynamic dateInput) {
    if (dateInput == null) return 'Recently';

    DateTime? date;
    if (dateInput is DateTime) {
      date = dateInput.toLocal();
    } else if (dateInput is String && dateInput.isNotEmpty) {
      date = DateTime.tryParse(dateInput)?.toLocal();
    }

    if (date == null) return 'Recently';

    final now = DateTime.now();
    final diff = now.difference(date);

    if (diff.isNegative) return 'Just now';

    final diffSecs = diff.inSeconds;
    final diffMins = diff.inMinutes;
    final diffHours = diff.inHours;
    final diffDays = diff.inDays;

    if (diffSecs < 60) {
      return 'Just now';
    }
    if (diffMins < 60) {
      return '${diffMins}m ago';
    }
    if (diffHours < 24) {
      return '${diffHours}h ago';
    }
    if (diffDays == 1) {
      return 'Yesterday';
    }
    if (diffDays < 14) {
      return '${diffDays}d ago';
    }
    if (diffDays < 60) {
      final weeks = (diffDays / 7).floor();
      return '${weeks}w ago';
    }
    if (diffDays < 365) {
      final months = (diffDays / 30).floor();
      return '${months}mo ago';
    }

    return 'Posted ${DateFormat('MMM d, yyyy').format(date)}';
  }

  /// Formats UTC ISO string or DateTime to exact date and time (e.g. "Posted August 24, 2026 at 3:42 PM").
  static String formatExactDateTime(dynamic dateInput) {
    if (dateInput == null) return 'Posted recently';

    DateTime? date;
    if (dateInput is DateTime) {
      date = dateInput.toLocal();
    } else if (dateInput is String && dateInput.isNotEmpty) {
      date = DateTime.tryParse(dateInput)?.toLocal();
    }

    if (date == null) return 'Posted recently';

    final dateStr = DateFormat('MMMM d, yyyy').format(date);
    final timeStr = DateFormat('h:mm a').format(date);

    return 'Posted $dateStr at $timeStr';
  }
}
